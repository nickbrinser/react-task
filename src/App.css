.App {
  text-align: center;
}
.card-container {
  display: flex;
  flex-direction: column;
  background-color: rgb(42, 45, 50);
  border: 1px solid grey;
  border-radius: 5px;
  padding: 25px; 
  cursor: pointer;
  -moz-osx-font-smoothing: grayscale;
  backface-visibility: hidden;
  
}
.card-list {
  width: 90vw;
  margin: 0 auto;
  display: grid;
  grid-template-columns: minmax(0, 1fr) minmax(0, 1fr);
  grid-gap: 20px;
  padding-top: 50px;
}

.alert-message {
  width: 50%;  
  text-align: center;
  margin-left: auto;
  margin-right: auto;   
}

.graph-visual {
  width: 50%;  
  text-align: center;
  margin-left: auto;
  margin-right: auto;  
}
