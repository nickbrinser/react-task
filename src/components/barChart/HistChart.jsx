import React from 'react'
import {
  Histogram,
  BarSeries,
  XAxis,
  YAxis,
  withParentSize
} from '@data-ui/histogram'
import { chartTheme } from '@data-ui/theme'

const ResponsiveHistogram = withParentSize(
  ({ parentWidth, ...rest }) => (
    <Histogram
      width={parentWidth}
      height={parentWidth * 0.5}
      {...rest}
    />

  )

)

const HistChart = (props) => {
  const maxYValue = Math.max(...props.data.map(d => d.count))

  return (
    <ResponsiveHistogram
      theme={chartTheme}
      ariaLabel='Histogram'
      orientation='vertical'
      valueAccessor={datum => datum}
      binType='numeric'
      binCount={22}
      renderTooltip={({ event, datum, data, color }) => (
        <div>
          <strong style={{ color: chartTheme.colors.categories[3] }}>
            {datum.bin0} to {datum.bin1}
          </strong>
          <div>
            <strong>count: </strong>
            {datum.count}
          </div>
        </div>
      )}
    >
      <BarSeries
        stroke={chartTheme.colors.categories[3]}
        fillOpacity={0.15}
        fill='rgb(202, 230, 151)'
        rawData={props.data}
      />
      <XAxis color='rgb(202, 230, 151)' />
      <YAxis label='Count' yDomain={[0, maxYValue]} />
    </ResponsiveHistogram>
  )
}

export default HistChart
