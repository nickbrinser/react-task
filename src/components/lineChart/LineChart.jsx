import React from 'react'
import { Line } from 'react-chartjs-2'

const LineChart = props => <Line data={props.data} type='line' options={props.option} />

export default LineChart
