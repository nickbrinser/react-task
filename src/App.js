import React, { useState, useEffect } from 'react'
import socketIOClient from 'socket.io-client'
import LineChart from './components/lineChart/LineChart'
import HistChart from './components/barChart/HistChart'
import { InputAlert } from './components/inputAlert/InputAlert'
import { Message } from 'semantic-ui-react'
import { ReactComponent as Loading } from './assets/graph-visual.svg'

import './App.css'

const socket = socketIOClient()

function App () {
  const [response, setResponse] = useState(false)
  const [alertField, setAlert] = useState('')
  const [lineChartData, setLine] = useState({
    labels: [],
    datasets: [
      {
        label: 'Line Chart',
        backgrounColor: 'rgba(0,0,0,0)',
        borderColor: '#ffffff',
        pointBackgrounColor: 'rgb(202, 230, 151)',
        pointBorderColor: 'rgb(202, 230, 151)',
        borderWidth: '2',
        lineTension: 0.45,
        data: []
      }
    ]
  })
  const [barChartData, setBar] = useState({
    data: []
  })
  const [currentValue, setCurrent] = useState()

  useEffect(() => {
    socket.on('data', res => {
      const y = new Date(res.timestamp).toLocaleTimeString()
      const x = res.value
      setLine((prevState) => ({
        ...prevState,
        labels: [...prevState.labels, y],
        datasets: [{ ...prevState.datasets[0], data: [...prevState.datasets[0].data, x] }]
      }))

      setBar((prevState) => ({
        ...prevState,
        data: [...prevState.data, x]
      }))
      setCurrent(res.value)
      setResponse(true)
    })
    socket.on('disconnect', () => {
      socket.disconnect()
      setResponse(false)
      console.log('disconnected from server')
    })
  }, [])
  const handleChange = (e) => {
    setAlert({
      alertField: e.target.value
    })
  }
  return (

    <div style={{ textAlign: 'center' }}>
      {response ? (
        <div>
          <div className='card-list'>
            <div className='card-container'>
              <LineChart
                data={lineChartData}
              />
            </div>
            <div className='card-container'>
              <HistChart
                data={barChartData.data}
              />
            </div>
          </div>
          <InputAlert
            placeholder='Alert Threshold'
            handleChange={handleChange}
          />
          {currentValue > parseInt(alertField.alertField) && alertField !== '' ? (
            <div className='alert-message'>
              <Message color='red' size='huge'>
                {' '}
                Threshold Exceeded {alertField.alertfield}!
              </Message>
            </div>
          ) : null}
        </div>
      ) : (
        <Loading className='graph-visual' />
      )}
    </div>
  )
}

export default App
