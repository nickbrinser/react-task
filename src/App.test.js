import React from 'react'
import ReactDOM from 'react-dom'
import { shallow } from 'enzyme'
import SocketMock from 'socket.io-mock'
import App from './App'
import HistChart from './components/barChart/HistChart'
import LineChart from './components/lineChart/LineChart'


it('renders without crashing', () => {
  shallow(<App />)
})

describe('<HistChart /> with empty dataset', () => {
  const initialProps = {
    data: []
  }
  const container = shallow(<HistChart data={initialProps.data} />)
  it('should match the snapshot', () => {
    expect(container.html()).toMatchSnapshot()
  })
})

describe('<LineChart /> with empty dataset', () => {
  const initialProps = {
    labels: [],
    datasets: [
      {
        label: 'Line Chart',
        backgrounColor: 'rgba(0,0,0,0)',
        borderColor: '#ffffff',
        pointBackgrounColor: 'rgb(202, 230, 151)',
        pointBorderColor: 'rgb(202, 230, 151)',
        borderWidth: '2',
        lineTension: 0.45,
        data: []
      }
    ]
  }
  const container = shallow(<LineChart data={initialProps} />)
  it('should match the snapshot', () => {
    expect(container.html()).toMatchSnapshot()
  })
})

describe('fast and isolated socket tests', () => {
  it('Sockets should be able to talk to each other without a server', () => {
    const socket = new SocketMock()

    socket.on('message', (message) => {
      expect(message).toBe('Hello World')
    })
    socket.socketClient.emit('message', 'Hello World')
  })
})