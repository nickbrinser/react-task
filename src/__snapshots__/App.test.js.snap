// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HistChart /> with empty dataset should match the snapshot 1`] = `"<div style=\\"width:100%;height:100%\\"></div>"`;

exports[`<LineChart /> with empty dataset should match the snapshot 1`] = `"<canvas height=\\"150\\" width=\\"300\\"></canvas>"`;
